{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bb9f1acc-d267-45b3-a20f-2db78cfcc723",
   "metadata": {},
   "source": [
    "# Analyze BAM pileups\n",
    "\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3f9e2ad-e8c1-4d6f-b9f2-2c9302c91a29",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-24T18:49:40.592106Z",
     "iopub.status.busy": "2021-04-24T18:49:40.591658Z",
     "iopub.status.idle": "2021-04-24T18:49:41.455332Z",
     "shell.execute_reply": "2021-04-24T18:49:41.454118Z",
     "shell.execute_reply.started": "2021-04-24T18:49:40.592000Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f7ac0ba-1cfa-46b6-b430-236dc9a34b1f",
   "metadata": {},
   "source": [
    "Get key variables from `Snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2e5b35e9-89c2-4c47-bbeb-58a4f89e6e97",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-24T18:54:12.011456Z",
     "iopub.status.busy": "2021-04-24T18:54:12.011103Z",
     "iopub.status.idle": "2021-04-24T18:54:12.015759Z",
     "shell.execute_reply": "2021-04-24T18:54:12.014804Z",
     "shell.execute_reply.started": "2021-04-24T18:54:12.011424Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#pileup_file = snakemake.input.pileup\n",
    "#aligner_key_file = snakemake.input.aligner_key\n",
    "#genome_key_file = snakemake.input.genome_key\n",
    "#sample_key_file = snakemake.input.sample_key\n",
    "#chart_file = snakemake.output.chart\n",
    "#diffs_from_reference_file = snakemake.output.diffs_from_reference\n",
    "#consensus_min_coverage = snakemake.params.consensus_min_coverage\n",
    "#conensus_min_frac = snakemake.params.consensus_min_frac\n",
    "\n",
    "pileup_file='../results/pileup/merged_pileup.csv'\n",
    "aligner_key_file='../results/pileup/merged_aligner_key.csv'\n",
    "genome_key_file='../results/pileup/merged_genome_key.csv'\n",
    "sample_key_file='../results/pileup/merged_sample_key.csv'\n",
    "chart='../results/pileup/merged_interactive_pileup_chart.html'\n",
    "diffs_from_reference='../results/pileup/merged_diffs_from_reference.csv'\n",
    "consensus_min_coverage = 5  # require >= this coverage\n",
    "consensus_min_frac = 0.75"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9dc57e7-e8ce-4561-a159-68a9d320203f",
   "metadata": {},
   "source": [
    "Read the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "29dc99f0-3836-4bad-822d-7cdb5e43da4a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-24T18:54:13.455799Z",
     "iopub.status.busy": "2021-04-24T18:54:13.455333Z",
     "iopub.status.idle": "2021-04-24T18:54:14.445765Z",
     "shell.execute_reply": "2021-04-24T18:54:14.445037Z",
     "shell.execute_reply.started": "2021-04-24T18:54:13.455759Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>ref_nt</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>aligner</th>\n",
       "      <th>genome</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015575</th>\n",
       "      <td>29866</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015576</th>\n",
       "      <td>29867</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015577</th>\n",
       "      <td>29868</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015578</th>\n",
       "      <td>29869</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015579</th>\n",
       "      <td>29870</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1015580 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          site ref_nt   A  C  G   T  aligner  genome  sample\n",
       "0            1      A  35  0  1   0        1       1       1\n",
       "1            2      T   0  0  0  39        1       1       1\n",
       "2            3      T   0  0  0  41        1       1       1\n",
       "3            4      A  41  0  0   0        1       1       1\n",
       "4            5      A  36  0  0   1        1       1       1\n",
       "...        ...    ...  .. .. ..  ..      ...     ...     ...\n",
       "1015575  29866      A   0  0  0   0        2       1      17\n",
       "1015576  29867      T   0  0  0   0        2       1      17\n",
       "1015577  29868      G   0  0  0   0        2       1      17\n",
       "1015578  29869      A   0  0  0   0        2       1      17\n",
       "1015579  29870      C   0  0  0   0        2       1      17\n",
       "\n",
       "[1015580 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pileup = pd.read_csv(pileup_file)\n",
    "aligner_key = pd.read_csv(aligner_key_file)\n",
    "genome_key = pd.read_csv(genome_key_file)\n",
    "sample_key = pd.read_csv(sample_key_file)\n",
    "\n",
    "assert set(pileup['aligner']) == set(aligner_key['aligner'])\n",
    "assert set(pileup['genome']) == set(genome_key['genome'])\n",
    "assert set(pileup['sample']) == set(sample_key['sample'])\n",
    "\n",
    "assert len(pileup) == len(pileup.drop_duplicates())\n",
    "assert pileup['site'].nunique() == pileup['site'].max()\n",
    "\n",
    "pileup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "646ab59a-df7b-4e14-9734-98fb6a4a9c43",
   "metadata": {
    "tags": []
   },
   "source": [
    "Identify differences between reference and consensus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "44daeb8d-e8bb-4d86-ab7f-6e1dde2b90d2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-24T19:01:54.834035Z",
     "iopub.status.busy": "2021-04-24T19:01:54.833555Z",
     "iopub.status.idle": "2021-04-24T19:01:55.780961Z",
     "shell.execute_reply": "2021-04-24T19:01:55.780443Z",
     "shell.execute_reply.started": "2021-04-24T19:01:54.833992Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>ref_nt</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>aligner</th>\n",
       "      <th>genome</th>\n",
       "      <th>sample</th>\n",
       "      <th>depth</th>\n",
       "      <th>consensus</th>\n",
       "      <th>consensus_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12312</th>\n",
       "      <td>12313</td>\n",
       "      <td>T</td>\n",
       "      <td>226</td>\n",
       "      <td>273</td>\n",
       "      <td>57</td>\n",
       "      <td>1131</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1687</td>\n",
       "      <td>T</td>\n",
       "      <td>0.670421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20976</th>\n",
       "      <td>20977</td>\n",
       "      <td>G</td>\n",
       "      <td>60</td>\n",
       "      <td>53</td>\n",
       "      <td>421</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>569</td>\n",
       "      <td>G</td>\n",
       "      <td>0.739895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26523</th>\n",
       "      <td>26524</td>\n",
       "      <td>T</td>\n",
       "      <td>249</td>\n",
       "      <td>202</td>\n",
       "      <td>335</td>\n",
       "      <td>1113</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1899</td>\n",
       "      <td>T</td>\n",
       "      <td>0.586098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26524</th>\n",
       "      <td>26525</td>\n",
       "      <td>G</td>\n",
       "      <td>246</td>\n",
       "      <td>157</td>\n",
       "      <td>1378</td>\n",
       "      <td>119</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1900</td>\n",
       "      <td>G</td>\n",
       "      <td>0.725263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26525</th>\n",
       "      <td>26526</td>\n",
       "      <td>G</td>\n",
       "      <td>198</td>\n",
       "      <td>217</td>\n",
       "      <td>1184</td>\n",
       "      <td>140</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1739</td>\n",
       "      <td>G</td>\n",
       "      <td>0.680851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911573</th>\n",
       "      <td>15474</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>T</td>\n",
       "      <td>0.722222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913243</th>\n",
       "      <td>17144</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>923207</th>\n",
       "      <td>27108</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "      <td>0.555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980162</th>\n",
       "      <td>24323</td>\n",
       "      <td>A</td>\n",
       "      <td>197</td>\n",
       "      <td>126</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>324</td>\n",
       "      <td>A</td>\n",
       "      <td>0.608025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009743</th>\n",
       "      <td>24034</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>9539</td>\n",
       "      <td>0</td>\n",
       "      <td>30237</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>39779</td>\n",
       "      <td>T</td>\n",
       "      <td>0.760125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>312 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          site ref_nt    A     C     G      T  aligner  genome  sample  depth  \\\n",
       "12312    12313      T  226   273    57   1131        1       1       1   1687   \n",
       "20976    20977      G   60    53   421     35        1       1       1    569   \n",
       "26523    26524      T  249   202   335   1113        1       1       1   1899   \n",
       "26524    26525      G  246   157  1378    119        1       1       1   1900   \n",
       "26525    26526      G  198   217  1184    140        1       1       1   1739   \n",
       "...        ...    ...  ...   ...   ...    ...      ...     ...     ...    ...   \n",
       "911573   15474      T    3     2     0     13        2       1      14     18   \n",
       "913243   17144      G    2     0     5      0        2       1      14      7   \n",
       "923207   27108      T    0     0     4      5        2       1      14      9   \n",
       "980162   24323      A  197   126     1      0        2       1      16    324   \n",
       "1009743  24034      C    3  9539     0  30237        2       1      17  39779   \n",
       "\n",
       "        consensus  consensus_frac  \n",
       "12312           T        0.670421  \n",
       "20976           G        0.739895  \n",
       "26523           T        0.586098  \n",
       "26524           G        0.725263  \n",
       "26525           G        0.680851  \n",
       "...           ...             ...  \n",
       "911573          T        0.722222  \n",
       "913243          G        0.714286  \n",
       "923207          T        0.555556  \n",
       "980162          A        0.608025  \n",
       "1009743         T        0.760125  \n",
       "\n",
       "[312 rows x 12 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nts = ['A', 'C', 'G', 'T']\n",
    "\n",
    "diffs_from_reference = (\n",
    "    pileup\n",
    "    .assign(depth=lambda x: x[nts].sum(axis=1),\n",
    "            consensus=lambda x: x[nts].idxmax(axis=1),\n",
    "            consensus_frac=lambda x: x[nts].max(axis=1) / x['depth']\n",
    "            )\n",
    "    .query('depth >= @consensus_min_coverage')\n",
    "    .query('(ref_nt != consensus) or (consensus_frac < @consensus_min_frac)')\n",
    "    )\n",
    "\n",
    "diffs_from_reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4859b5c7-b45b-47e4-8e22-891b0d4964d7",
   "metadata": {},
   "source": [
    "Make interactive pilup plot (this may be very large, and won't scale well if many more samples are added):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c9dd374-35a1-4d81-8a0c-b8e68e715cfe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "width = 900\n",
    "\n",
    "selections = {}\n",
    "for key_df, field, name in [(aligner_key, 'aligner', 'read'),\n",
    "                            (genome_key, 'genome', 'reference'),\n",
    "                            (sample_key, 'sample', 'viral'),\n",
    "                            ]:\n",
    "    selections[field] = alt.selection_single(\n",
    "            name=name,\n",
    "            fields=[field],\n",
    "            bind=alt.binding_select(options=key_df[field].tolist(),\n",
    "                                    labels=key_df[field + '_name'].tolist()),\n",
    "            init={field: key_df[field].tolist()[0]}\n",
    "            )\n",
    "\n",
    "zoom_brush = alt.selection_interval(\n",
    "                encodings=['x'],\n",
    "                mark=alt.BrushConfig(stroke='black', strokeWidth=2)\n",
    "                )\n",
    "    \n",
    "site_zoom = (\n",
    "    alt.Chart(pileup[['site']].drop_duplicates())\n",
    "    .mark_rect(color='lightgray')\n",
    "    .encode(x=alt.X('site:O',\n",
    "                    axis=alt.Axis(grid=False,\n",
    "                                  ticks=False,\n",
    "                                  ),\n",
    "                    title=None)\n",
    "            )\n",
    "    .add_selection(zoom_brush)\n",
    "    .properties(width=width,\n",
    "                height=15,\n",
    "                title='site zoom bar')\n",
    "    )\n",
    "\n",
    "pileup_chart = (\n",
    "    alt.Chart(pileup)\n",
    "    .encode(x=alt.X('site:O',\n",
    "                    axis=alt.Axis(grid=False,\n",
    "                                  ticks=False,\n",
    "                                  ),\n",
    "                    ),\n",
    "            y='depth:Q',\n",
    "            tooltip=['site', 'depth:Q', 'A', 'C', 'G', 'T',\n",
    "                     alt.Tooltip('ref_nt', title='reference')],\n",
    "            )\n",
    "    .mark_bar()\n",
    "    .add_selection(selections['aligner'],\n",
    "                   selections['genome'],\n",
    "                   selections['sample'],\n",
    "                   zoom_brush)\n",
    "    .transform_filter(selections['aligner'])\n",
    "    .transform_filter(selections['genome'])\n",
    "    .transform_filter(selections['sample'])\n",
    "    .transform_filter(zoom_brush)\n",
    "    .transform_calculate(\n",
    "        depth=alt.datum.A + alt.datum.C + alt.datum.G + alt.datum.T\n",
    "        )\n",
    "    .properties(height=250,\n",
    "                width=width,\n",
    "                title='read coverage over genome',\n",
    "                )\n",
    "    )\n",
    "\n",
    "zoomable_pileup_chart = site_zoom & pileup_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fff961d-d77b-474a-954d-50a711b89da1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
