{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c7705cb-5a89-4ec4-b0cd-2fa78026b255",
   "metadata": {},
   "source": [
    "# Analyze consensus from deep sequencing pileup versus Genbank accessions\n",
    "\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "23c2d7c5-db4b-43f9-a677-4124a121e859",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-22T16:25:24.633454Z",
     "iopub.status.busy": "2021-04-22T16:25:24.632683Z",
     "iopub.status.idle": "2021-04-22T16:25:24.814637Z",
     "shell.execute_reply": "2021-04-22T16:25:24.813650Z",
     "shell.execute_reply.started": "2021-04-22T16:25:24.633389Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import Bio.AlignIO\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f443a738-19fb-47f1-8cd8-6f23011002dd",
   "metadata": {},
   "source": [
    "Get the pileup-consensus / Genbank alignments along with the descriptions from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1d5fb0b6-aff0-45f2-907e-ddb9721deea3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-22T16:26:11.243336Z",
     "iopub.status.busy": "2021-04-22T16:26:11.242701Z",
     "iopub.status.idle": "2021-04-22T16:26:11.257778Z",
     "shell.execute_reply": "2021-04-22T16:26:11.257117Z",
     "shell.execute_reply.started": "2021-04-22T16:26:11.243290Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#alignments = snakemake.input.alignments\n",
    "#descriptors = snakemake.params.descriptors\n",
    "\n",
    "alignments = ['results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR10971381.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092056.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092057.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092058.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092059.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092060.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092061.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092062.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092063.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11092064.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR10903401.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR10903402.fa',\n",
    " 'results/consensus_to_genbank_alignments/bbmap/SARS-CoV-2/SRR11177792.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR10971381.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092056.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092057.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092058.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092059.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092060.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092061.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092062.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092063.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11092064.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR10903401.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR10903402.fa',\n",
    " 'results/consensus_to_genbank_alignments/bwa-mem2/SARS-CoV-2/SRR11177792.fa']\n",
    "alignments = ['../' + a for a in alignments]\n",
    "\n",
    "descriptors = [{'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR10971381'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092056'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092057'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092058'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092059'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092060'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092061'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092062'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092063'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092064'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR10903401'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR10903402'},\n",
    " {'aligner': 'bbmap', 'genome': 'SARS-CoV-2', 'accession': 'SRR11177792'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR10971381'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092056'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092057'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092058'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092059'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092060'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092061'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092062'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092063'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11092064'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR10903401'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR10903402'},\n",
    " {'aligner': 'bwa-mem2', 'genome': 'SARS-CoV-2', 'accession': 'SRR11177792'}]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e531a8d-87f2-45b4-b9c1-f6c2ede8003b",
   "metadata": {},
   "source": [
    "Create a data frame with the alignments, and then process it to get:\n",
    " - total sites in alignment\n",
    " - sites that are identical\n",
    " - sites that are ambiguous (a `N` in one sequence)\n",
    " - sites that are gapped in either sequence but have a called identity in other\n",
    " - sites that are mismatched called nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6187c520-943f-435d-85f3-1d4303da9d31",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-22T17:27:24.992146Z",
     "iopub.status.busy": "2021-04-22T17:27:24.991791Z",
     "iopub.status.idle": "2021-04-22T17:27:25.133013Z",
     "shell.execute_reply": "2021-04-22T17:27:25.132458Z",
     "shell.execute_reply.started": "2021-04-22T17:27:24.992115Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>aligner</th>\n",
       "      <th>genome</th>\n",
       "      <th>alignment_file</th>\n",
       "      <th>total</th>\n",
       "      <th>identical</th>\n",
       "      <th>ambiguous</th>\n",
       "      <th>mismatched</th>\n",
       "      <th>gapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10971381</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29903</td>\n",
       "      <td>29870</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11092056</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>1871</td>\n",
       "      <td>27999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11092057</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29891</td>\n",
       "      <td>29630</td>\n",
       "      <td>240</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11092058</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>9581</td>\n",
       "      <td>20288</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11092059</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>29746</td>\n",
       "      <td>124</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR11092060</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>27033</td>\n",
       "      <td>2837</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR11092061</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>24589</td>\n",
       "      <td>5280</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR11092062</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29891</td>\n",
       "      <td>29857</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR11092063</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>25106</td>\n",
       "      <td>4752</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR11092064</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>16522</td>\n",
       "      <td>13347</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR10903401</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29882</td>\n",
       "      <td>29815</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR10903402</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29882</td>\n",
       "      <td>29811</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>bbmap</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bbm...</td>\n",
       "      <td>29870</td>\n",
       "      <td>29811</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR10971381</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29903</td>\n",
       "      <td>29870</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR11092056</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>1891</td>\n",
       "      <td>27979</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR11092057</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29891</td>\n",
       "      <td>29630</td>\n",
       "      <td>240</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR11092058</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>9581</td>\n",
       "      <td>20288</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR11092059</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>29748</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR11092060</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>27011</td>\n",
       "      <td>2858</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR11092061</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>24558</td>\n",
       "      <td>5311</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR11092062</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29891</td>\n",
       "      <td>29857</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR11092063</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>25098</td>\n",
       "      <td>4749</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR11092064</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>16623</td>\n",
       "      <td>13246</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR10903401</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29882</td>\n",
       "      <td>29869</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SRR10903402</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29882</td>\n",
       "      <td>29848</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>bwa-mem2</td>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>../results/consensus_to_genbank_alignments/bwa...</td>\n",
       "      <td>29870</td>\n",
       "      <td>29811</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession   aligner      genome  \\\n",
       "0   SRR10971381     bbmap  SARS-CoV-2   \n",
       "1   SRR11092056     bbmap  SARS-CoV-2   \n",
       "2   SRR11092057     bbmap  SARS-CoV-2   \n",
       "3   SRR11092058     bbmap  SARS-CoV-2   \n",
       "4   SRR11092059     bbmap  SARS-CoV-2   \n",
       "5   SRR11092060     bbmap  SARS-CoV-2   \n",
       "6   SRR11092061     bbmap  SARS-CoV-2   \n",
       "7   SRR11092062     bbmap  SARS-CoV-2   \n",
       "8   SRR11092063     bbmap  SARS-CoV-2   \n",
       "9   SRR11092064     bbmap  SARS-CoV-2   \n",
       "10  SRR10903401     bbmap  SARS-CoV-2   \n",
       "11  SRR10903402     bbmap  SARS-CoV-2   \n",
       "12  SRR11177792     bbmap  SARS-CoV-2   \n",
       "13  SRR10971381  bwa-mem2  SARS-CoV-2   \n",
       "14  SRR11092056  bwa-mem2  SARS-CoV-2   \n",
       "15  SRR11092057  bwa-mem2  SARS-CoV-2   \n",
       "16  SRR11092058  bwa-mem2  SARS-CoV-2   \n",
       "17  SRR11092059  bwa-mem2  SARS-CoV-2   \n",
       "18  SRR11092060  bwa-mem2  SARS-CoV-2   \n",
       "19  SRR11092061  bwa-mem2  SARS-CoV-2   \n",
       "20  SRR11092062  bwa-mem2  SARS-CoV-2   \n",
       "21  SRR11092063  bwa-mem2  SARS-CoV-2   \n",
       "22  SRR11092064  bwa-mem2  SARS-CoV-2   \n",
       "23  SRR10903401  bwa-mem2  SARS-CoV-2   \n",
       "24  SRR10903402  bwa-mem2  SARS-CoV-2   \n",
       "25  SRR11177792  bwa-mem2  SARS-CoV-2   \n",
       "\n",
       "                                       alignment_file  total  identical  \\\n",
       "0   ../results/consensus_to_genbank_alignments/bbm...  29903      29870   \n",
       "1   ../results/consensus_to_genbank_alignments/bbm...  29870       1871   \n",
       "2   ../results/consensus_to_genbank_alignments/bbm...  29891      29630   \n",
       "3   ../results/consensus_to_genbank_alignments/bbm...  29870       9581   \n",
       "4   ../results/consensus_to_genbank_alignments/bbm...  29870      29746   \n",
       "5   ../results/consensus_to_genbank_alignments/bbm...  29870      27033   \n",
       "6   ../results/consensus_to_genbank_alignments/bbm...  29870      24589   \n",
       "7   ../results/consensus_to_genbank_alignments/bbm...  29891      29857   \n",
       "8   ../results/consensus_to_genbank_alignments/bbm...  29870      25106   \n",
       "9   ../results/consensus_to_genbank_alignments/bbm...  29870      16522   \n",
       "10  ../results/consensus_to_genbank_alignments/bbm...  29882      29815   \n",
       "11  ../results/consensus_to_genbank_alignments/bbm...  29882      29811   \n",
       "12  ../results/consensus_to_genbank_alignments/bbm...  29870      29811   \n",
       "13  ../results/consensus_to_genbank_alignments/bwa...  29903      29870   \n",
       "14  ../results/consensus_to_genbank_alignments/bwa...  29870       1891   \n",
       "15  ../results/consensus_to_genbank_alignments/bwa...  29891      29630   \n",
       "16  ../results/consensus_to_genbank_alignments/bwa...  29870       9581   \n",
       "17  ../results/consensus_to_genbank_alignments/bwa...  29870      29748   \n",
       "18  ../results/consensus_to_genbank_alignments/bwa...  29870      27011   \n",
       "19  ../results/consensus_to_genbank_alignments/bwa...  29870      24558   \n",
       "20  ../results/consensus_to_genbank_alignments/bwa...  29891      29857   \n",
       "21  ../results/consensus_to_genbank_alignments/bwa...  29870      25098   \n",
       "22  ../results/consensus_to_genbank_alignments/bwa...  29870      16623   \n",
       "23  ../results/consensus_to_genbank_alignments/bwa...  29882      29869   \n",
       "24  ../results/consensus_to_genbank_alignments/bwa...  29882      29848   \n",
       "25  ../results/consensus_to_genbank_alignments/bwa...  29870      29811   \n",
       "\n",
       "    ambiguous  mismatched  gapped  \n",
       "0           0           0      33  \n",
       "1       27999           0       0  \n",
       "2         240           0      21  \n",
       "3       20288           1       0  \n",
       "4         124           0       0  \n",
       "5        2837           0       0  \n",
       "6        5280           1       0  \n",
       "7          13           0      21  \n",
       "8        4752           1      11  \n",
       "9       13347           1       0  \n",
       "10         55           0      12  \n",
       "11         59           0      12  \n",
       "12         25           0      34  \n",
       "13          0           0      33  \n",
       "14      27979           0       0  \n",
       "15        240           0      21  \n",
       "16      20288           1       0  \n",
       "17        122           0       0  \n",
       "18       2858           1       0  \n",
       "19       5311           1       0  \n",
       "20         13           0      21  \n",
       "21       4749           1      22  \n",
       "22      13246           1       0  \n",
       "23          0           0      13  \n",
       "24         22           0      12  \n",
       "25         35           0      24  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assert len(alignments) == len(descriptors)\n",
    "\n",
    "alignments_df = (\n",
    "    pd.DataFrame({'accession': [d['accession'] for d in descriptors],\n",
    "                  'aligner': [d['aligner'] for d in descriptors],\n",
    "                  'genome': [d['genome'] for d in descriptors],\n",
    "                  'alignment_file': alignments,\n",
    "                  })\n",
    "    )\n",
    "\n",
    "def alignment_stats(aln_file):\n",
    "    aln = Bio.AlignIO.read(aln_file, 'fasta')\n",
    "    assert len(aln) == 2\n",
    "    s1 = str(aln[0].seq).upper()\n",
    "    s2 = str(aln[1].seq).upper()\n",
    "    assert len(s1) == len(s2)\n",
    "    total_sites = len(s1)\n",
    "    identical_sites = ambiguous_sites = mismatched_sites = gapped_sites = 0\n",
    "    for nt1, nt2 in zip(s1, s2):\n",
    "        if nt1 == nt2:\n",
    "            identical_sites += 1\n",
    "        elif nt1 == 'N' or nt2 == 'N':\n",
    "            ambiguous_sites += 1\n",
    "        elif nt1 == '-' or nt2 == '-':\n",
    "            gapped_sites += 1\n",
    "        else:\n",
    "            mismatched_sites += 1\n",
    "    assert total_sites == identical_sites + ambiguous_sites + mismatched_sites + gapped_sites\n",
    "    return total_sites, identical_sites, ambiguous_sites, mismatched_sites, gapped_sites\n",
    "\n",
    "for col, vals in zip(['total', 'identical', 'ambiguous', 'mismatched', 'gapped'],\n",
    "                     zip(*alignments_df['alignment_file'].map(alignment_stats))\n",
    "                     ):\n",
    "    alignments_df[col] = vals\n",
    "\n",
    "alignment_stats(alignments_df.at[0, 'alignment_file'])\n",
    "\n",
    "alignments_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "094d06f4-adc3-4cbb-a25b-3cbd69ef9223",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-04-22T16:40:30.775661Z",
     "iopub.status.busy": "2021-04-22T16:40:30.775180Z",
     "iopub.status.idle": "2021-04-22T16:40:30.782552Z",
     "shell.execute_reply": "2021-04-22T16:40:30.781515Z",
     "shell.execute_reply.started": "2021-04-22T16:40:30.775620Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('a', 'c'), ('b', 'd'), ('e', 'f')]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(zip(*[('a', 'b', 'e'), ('c', 'd', 'f')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3962239d-f949-46fc-9bb1-a9350e91c719",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
